### PR TITLE
JENKINS-53794 JENKINS-56987 Bump to jira rest client 5.2.1

### PR DESCRIPTION
**Pull request by rodrigc**
JENKINS-53794 JENKINS-56987 Bump to jira rest client 5.2.0

This is taken from: https://github.com/jenkinsci/jira-plugin/pull/212

See also:
*  https://bitbucket.org/atlassian/jira-rest-java-client/pull-requests/93 
*  https://bitbucket.org/atlassian/jira-rest-java-client/pull-requests/93/jrjc-240-added-support-for-accountid/d

### CODE REVIEW

**Comment by olamy** [comment]
@rodrigc thanks but there is already the pr #212 this is a bit confusing to have another one..

**Comment by rodrigc** [comment]
@olamy #212 does not fully compile or pass tests, so I am trying to work on that in this PR, and am trying to get something I can test today, since I need this.  I can combine efforts.

**Comment by rodrigc** [comment]
@olamy @ca-stefan-cordes @warden I built a jira-plugin off of this branch, and managed to get it to work against JIRA Cloud with the latest GDPR changes.
The only thing is that I cannot get the tests to succeed because it is missing a dependency on the Google guava library.  I am not enough of a POM expert to fix that.

Can you help?

@ca-stefan-cordes thanks for your initial patch, I used that to move forward.

I need to get this to work because I use the jira-plugin very heavily against the JIRA Cloud plugin at my work.

**Comment by olamy** [comment]
why 7a524c87a1a531c7433fd265fbc78183c9e9a5de ???
we did it on purpose to control what is happening here....

**Comment by rodrigc** [comment]
I did https://github.com/jenkinsci/jira-plugin/commit/7a524c87a1a531c7433fd265fbc78183c9e9a5de because the plugin didn't seem to work without it after pulling in the new dependencies

**Comment by rodrigc** [comment]
@olamy In this branch, how can I get rid of this error which occurs during the tests:

```
java.lang.NoClassDefFoundError: com/google/common/base/MoreObjects$ToStringHelper
```

That is due to dependency on guava.  I tried playing with the pom.xml but couldn't get rid of it.

**Comment by olamy** [comment]
would be better to search for the reasons of this "didn't seem to work" we did that to be able to control proxy setup. 
It means with removing this you broke jenkins proxy support and maybe other cases....

**Comment by rodrigc** [comment]
In my setup, with this branch, I've been able to get things to create an issue against JIRA Cloud, after the GDPR server changes.  The only minor problem I am facing is that even though I populate the *assignee* field, the issue created in JIRA cloud shows up as *Unassigned*.  So that looks like part of the API change made by Atlassian.

I'm going to focus on fixing that in this branch.

If @olamy and @ca-stefan-cordes want to collaborate on fixing the httpclient in this branch that would be great.

The past 3 days, I have been pushing Atlassian to fix their rest client and push out a new version, which thankfully they did.

**Comment by rodrigc** [comment]
OK, I fixed the following:

1.  guava dependency: https://github.com/jenkinsci/jira-plugin/pull/213/commits/d39e7892304f255a9bcd33e2da06fc2de9f0ba85
2. "assignee" was not getting populated: https://github.com/jenkinsci/jira-plugin/pull/213/commits/4a6a1c5aa991172eca02131cdcf14595ee0e97f0

I'm running this patch now in my setup and it is working against JIRA Cloud.

When trying to run `mvn test` I am still seeing some nullpoint exceptions, but haven't tracked those down yet.

**Comment by rodrigc** [comment]
I'm using Jenkins LTS 2.204.2

**Comment by rodrigc** [comment]
For the most part this patch seems to be holding up in my environment.  I can successfully create new issues on job failure, and even update them.

A few times, I am seeing this:

```
2020-02-24 21:10:58.189+0000 [id=141]	WARNING	h.plugins.jira.JiraRestService#getIssue: jira rest client get issue error. cause: RestClientException{statusCode=Optional.of(405), errorCollections=[]}
org.codehaus.jettison.json.JSONException: A JSONObject text must begin with '{' at character 0 of
	at org.codehaus.jettison.json.JSONTokener.syntaxError(JSONTokener.java:439)
	at org.codehaus.jettison.json.JSONObject.<init>(JSONObject.java:169)
	at org.codehaus.jettison.json.JSONObject.<init>(JSONObject.java:266)
	at com.atlassian.jira.rest.client.internal.async.AbstractAsynchronousRestClient.extractErrors(AbstractAsynchronousRestClient.java:212)
	at com.atlassian.jira.rest.client.internal.async.AbstractAsynchronousRestClient$2.apply(AbstractAsynchronousRestClient.java:175)
Caused: RestClientException{statusCode=Optional.of(405), errorCollections=[]}
	at com.atlassian.jira.rest.client.internal.async.AbstractAsynchronousRestClient$2.apply(AbstractAsynchronousRestClient.java:178)
	at com.atlassian.jira.rest.client.internal.async.AbstractAsynchronousRestClient$2.apply(AbstractAsynchronousRestClient.java:170)
	at com.atlassian.httpclient.api.ResponsePromiseMapFunction.apply(ResponsePromiseMapFunction.java:49)
	at com.atlassian.httpclient.api.ResponsePromiseMapFunction.apply(ResponsePromiseMapFunction.java:10)
	at io.atlassian.util.concurrent.Promises$OfStage.lambda$fold$4(Promises.java:332)
	at io.atlassian.util.concurrent.Promises.lambda$biFunction$7(Promises.java:422)
	at java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:836)
	at java.util.concurrent.CompletableFuture$UniHandle.tryFire(CompletableFuture.java:811)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:488)
	at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1975)
	at com.atlassian.httpclient.apache.httpcomponents.SettableFuturePromiseHttpPromiseAsyncClient$1.lambda$doCompleted$0(SettableFuturePromiseHttpPromiseAsyncClient.java:37)
	at com.atlassian.httpclient.apache.httpcomponents.SettableFuturePromiseHttpPromiseAsyncClient.runInContext(SettableFuturePromiseHttpPromiseAsyncClient.java:61)
	at com.atlassian.httpclient.apache.httpcomponents.SettableFuturePromiseHttpPromiseAsyncClient$ThreadLocalDelegateRunnable.run(SettableFuturePromiseHttpPromiseAsyncClient.java:129)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused: java.util.concurrent.ExecutionException
	at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:357)
	at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1928)
	at io.atlassian.util.concurrent.Promises$OfStage.get(Promises.java:357)
	at com.atlassian.jira.rest.client.internal.async.DelegatingPromise.get(DelegatingPromise.java:106)
	at hudson.plugins.jira.JiraRestService.getIssue(JiraRestService.java:156)
	at hudson.plugins.jira.JiraSession.getIssueByKey(JiraSession.java:433)
	at hudson.plugins.jira.JiraCreateIssueNotifier.getStatus(JiraCreateIssueNotifier.java:219)
	at hudson.plugins.jira.JiraCreateIssueNotifier.currentBuildResultSuccess(JiraCreateIssueNotifier.java:388)
	at hudson.plugins.jira.JiraCreateIssueNotifier.perform(JiraCreateIssueNotifier.java:161)
	at hudson.tasks.BuildStepMonitor$1.perform(BuildStepMonitor.java:20)
	at hudson.model.AbstractBuild$AbstractBuildExecution.perform(AbstractBuild.java:741)
	at hudson.model.AbstractBuild$AbstractBuildExecution.performAllBuildSteps(AbstractBuild.java:690)
	at hudson.model.Build$BuildExecution.post2(Build.java:186)
	at hudson.model.AbstractBuild$AbstractBuildExecution.post(AbstractBuild.java:635)
	at hudson.model.Run.execute(Run.java:1878)
	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
	at hudson.model.ResourceController.execute(ResourceController.java:97)
	at hudson.model.Executor.run(Executor.java:427)
```

I'm not sure what causes the HTTP 405 error.

**Comment by rodrigc** [comment]
Found this also:

```
java.lang.NullPointerException
	at hudson.plugins.jira.JiraCreateIssueNotifier.getStatus(JiraCreateIssueNotifier.java:220)
	at hudson.plugins.jira.JiraCreateIssueNotifier.currentBuildResultSuccess(JiraCreateIssueNotifier.java:388)
	at hudson.plugins.jira.JiraCreateIssueNotifier.perform(JiraCreateIssueNotifier.java:161)
	at hudson.tasks.BuildStepMonitor$1.perform(BuildStepMonitor.java:20)
	at hudson.model.AbstractBuild$AbstractBuildExecution.perform(AbstractBuild.java:741)
	at hudson.model.AbstractBuild$AbstractBuildExecution.performAllBuildSteps(AbstractBuild.java:690)
	at hudson.model.Build$BuildExecution.post2(Build.java:186)
	at hudson.model.AbstractBuild$AbstractBuildExecution.post(AbstractBuild.java:635)
	at hudson.model.Run.execute(Run.java:1878)
	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
	at hudson.model.ResourceController.execute(ResourceController.java:97)
	at hudson.model.Executor.run(Executor.java:427)
```

**Comment by ca-stefan-cordes** [comment]
Hi @rodrigc ,
removing the (previously patched: [JENKINS-48357] get rid of atlassian http client (#145) ) 
com.atlassian.httpclient.apache.httpcomponents.*
is not a good idea to combine it with the jira-rest-client update.
(That's probably the reason why the tests are failing).

Better is to smoothly update to jira-rest-client.version_5.2.0 via updating the existing source files and creating another pull-request which reverts JENKINS-48357 (where they probably had good reasons to copy the sources) or to update/merge the copied com.atlassian.httpclient.apache.httpcomponents.*
with the actual one.

According to "Add guava dependency"  (commit d39e7892304f255a9bcd33e2da06fc2de9f0ba85) it would be better to remove the exclusion 5 lines above to simply use the transitive dependency (as jira-rest-java-client-api requires guava 26):
```
    <dependency>
      <groupId>com.atlassian.jira</groupId>
      <artifactId>jira-rest-java-client-api</artifactId>
      <version>${jira-rest-client.version}</version>
      <exclusions>
        <exclusion>
          <groupId>com.google.guava</groupId>
          <artifactId>guava</artifactId>
        </exclusion>
      </exclusions>
    </dependency>
    <dependency>
      <groupId>com.google.guava</groupId>
      <artifactId>guava</artifactId>
      <version>19.0</version>
    </dependency>
```
=>
```
    <dependency>
      <groupId>com.atlassian.jira</groupId>
      <artifactId>jira-rest-java-client-api</artifactId>
      <version>${jira-rest-client.version}</version>
    </dependency>
```

**Comment by rodrigc** [comment]
@ca-stefan-cordes thanks for your work, and feedback!  I spent a lot of time with your initial patch, and testing in my environment over the weekend.  If you want to collaborate more, we can do it here in the GitHub audit trail, or on Gitter https://gitter.im/jenkinsci/jira-plugin

**Comment by rodrigc** [comment]
BTW, I found a problem in this function:

```
 private String getIssue(String filename) throws IOException, InterruptedException {
        String issueId = "";
        Path path = Paths.get(filename);
        if(Files.notExists(path)){
            return null;
        }
        try (BufferedReader br = Files.newBufferedReader(path)) {
            String issue;

            while ((issue = br.readLine()) != null) {
                issueId = issue;
            }
            return issueId;
        } catch (FileNotFoundException e) {
            return null;
        }
    }
```

For whatever reason, I found a bunch of  jobs on my Jenkins server where `builds/issue.txt` was an empty file.  This might have been caused when Atlassian switched their API, and the JIRA plugin started failing in a bunch of places, causing empty issue.txt files to be written out.

If this file is empty, it causes error paths in other parts of the plugin.
The consequence of this is that the builds with the empty issue.txt files can never transition from failure to success, because trying to update the issue will always fail.

**Comment by rodrigc** [comment]
@ca-stefan-cordes if I remove the guava dependency as you suggest, then I get a compilation failure:

```
[INFO] -------------------------------------------------------------
[ERROR] COMPILATION ERROR :
[INFO] -------------------------------------------------------------
[ERROR] /Users/craigrodrigues/jira-plugin/src/main/java/hudson/plugins/jira/JiraSite.java:[756,23] cannot find symbol
  symbol:   method firstNonNull(java.net.URL,java.net.URL)
  location: class com.google.common.base.Objects
[ERROR] /Users/craigrodrigues/jira-plugin/src/main/java/hudson/plugins/jira/extension/ExtendedVersionInput.java:[19,31] cannot find symbol
  symbol:   method toStringHelper(hudson.plugins.jira.extension.ExtendedVersionInput)
  location: class com.google.common.base.Objects
[INFO] 2 errors
```

For some reason I found that moving to guava 26 would give this error, but using guava 19 would still work.

**Comment by ca-stefan-cordes** [comment]
@rodrigc : toStringHelper was copied to MoreObjects since guava 18.
The methods in guava Objects were removed with [version 21:](https://github.com/google/guava/commit/f67ab864bde63fa6602b5688de0440957ffeaa2e#diff-4bc2227e6f3230ff3b92daa36c6d8eec).

According to Guava I missed that jenkins provides 11.0.1
(where your original problem: java.lang.NoClassDefFoundError: com/google/common/base/MoreObjects$ToStringHelper came from).

So I need to correct myself. You have 2 options:
1) Keep exclusion and add Guava 18 to avoid `NoClassDefFoundError com/google/common/base/MoreObjects$ToStringHelper` (maybe add 19 or 20)
2) Remove Guava dependency (and so take 26.0-jre version)
 and replace Objects.toStringHelpers with MoreObjects.toStringHelper (same for firstNonNull)
 and hope that no methods are using Objects.ToStringHelper within stapler or guice:
     image

To step forward and not opening too many "construction sites" I would proceed with "1)".

**Comment by ca-stefan-cordes** [comment]
Oops, just see it is in my PR #212 which causes the
```
[2020-02-25T10:12:16.997Z] [ERROR] callGetIssuesFromJqlSearch(hudson.plugins.jira.selector.JqlIssueSelectorTest)  Time elapsed: 0 s  <<< ERROR!
[2020-02-25T10:12:16.997Z] java.lang.NoClassDefFoundError: com/google/common/base/MoreObjects$ToStringHelper
```
as the com.atlassian.jira.rest.client.api.domain.Issue is overriding
```
   @Override
    protected MoreObjects.ToStringHelper getToStringHelper() {
```

I'll try there.

**Comment by rodrigc** [comment]
@ca-stefan-cordes  Thanks for your analysis!  I would not have figured this stuff out...
  Let me know what you come up with

**Comment by ca-stefan-cordes** [comment]
@rodrigc : My build #213 is green now. You may merge my changes.

Together with the guava dependency there needed to be some more changed due to 
`Promises.forListenableFuture`
changes to 
`Promises.forFuture`
in the atlassian-util-concurrent library 3.0.0 => 4.0.0/.1

**Comment by rodrigc** [comment]
@ca-stefan-cordes OK, I backed out https://github.com/jenkinsci/jira-plugin/commit/7a524c87a1a531c7433fd265fbc78183c9e9a5de and merged your changes into this branch.  I am testing this now against JIRA Cloud and it seems to work.

Can you look?

Also, can you look at  https://github.com/jenkinsci/jira-plugin/pull/214 ?
I found that when I had a bunch of zero-length issue.txt files, things were in bad shape.

**Comment by ca-stefan-cordes** [comment]
The failing test
ApacheAsyncHttpClientTest.simple_get_with_non_proxy_host
looks like to work now (maybe temporary problem with apache host).

The test
JiraCreateIssueNotifierTest.performFailureSuccessIssueClosedWithComponents:192 expected:<0> but was:<1>
still fails - I'll take a look into it.

**Comment by rodrigc** [comment]
@ca-stefan-cordes Even though the test fails, I'm running this patch on my setup, and at least it seems to work against the GDPR changes in Atlassian cloud

**Comment by rodrigc** [comment]
@ca-stefan-cordes can you also take a look at https://bitbucket.org/atlassian/jira-rest-java-client/pull-requests/104/change-field-name-from-name-to-id-for/diff ?  I think we need to get Atlassian to merge that, otherwise populating the Assignee field doesn't work after the GDPR changes.

**Comment by ca-stefan-cordes** [comment]
setFieldInput in JiraRestService looks like to be feasible workaround until jira-rest-java-client is merged.

But shouldn't be the field name "accountId" instead of "id" ?

(again my build was fine https://travis-ci.com/c-a-services/jira-plugin/ - according to JiraCreateIssueNotifierTest (which is green locally, too)).

File: src/main/java/hudson/plugins/jira/JiraRestService.java [hunk 1]
```diff
@@ -275,8 +277,15 @@ public BasicIssue createIssue(String projectKey, String description, String assi
             builder.setPriorityId(priorityId);
         }
 
-        if (!assignee.equals(""))
-            builder.setAssigneeName(assignee);
+        if (!assignee.equals("")) {
```
Comment (olamy) [inline src/main/java/hudson/plugins/jira/JiraRestService.java]: please use StringUtils.isBlank here to avoid NPE

File: pom.xml [hunk 2]
```diff
@@ -21,8 +21,8 @@
     <java.level>8</java.level>
     <mockito.version>1.10.19</mockito.version>
     <powermock.version>1.6.3</powermock.version>
-    <jira-rest-client.version>5.0.4</jira-rest-client.version>
-    <fugue.version>2.6.1</fugue.version>
+    <jira-rest-client.version>5.2.0</jira-rest-client.version>
```
Comment (olamy) [inline pom.xml]: 5.2.1 looks to be available now

File: src/test/java/hudson/plugins/jira/JiraCreateIssueNotifierTest.java [hunk 3]
```diff
@@ -168,7 +169,7 @@ public void performFailureSuccessIssueClosedWithComponents() throws Exception {
         doReturn(site).when(notifier).getSiteForProject(Mockito.any());
 
         Issue issue = mock(Issue.class);
-        Status status = new Status(null, null, JiraCreateIssueNotifier.finishedStatuses.Closed.toString() , null, null);
+        Status status = new Status(null, null, "", JiraCreateIssueNotifier.finishedStatuses.Closed.toString() , null, null);
```
Comment (olamy) [inline src/test/java/hudson/plugins/jira/JiraCreateIssueNotifierTest.java]: this is the reason of the test failure. Status name must cannot be empty in the test.

File: pom.xml (hunk 2 above)
Comment (rodrigc) [reply pom.xml]: Bumping up to 5.2.1

File: src/main/java/hudson/plugins/jira/JiraRestService.java (hunk 1 above)
Comment (rodrigc) [reply src/main/java/hudson/plugins/jira/JiraRestService.java]: Fixed

**Comment by rodrigc** [comment]
@ca-stefan-cordes I changed to use *accountId*

File: src/test/java/hudson/plugins/jira/JiraCreateIssueNotifierTest.java (hunk 3 above)
Comment (rodrigc) [reply src/test/java/hudson/plugins/jira/JiraCreateIssueNotifierTest.java]: I changed this to have a name which is not empty.

**Comment by rodrigc** [comment]
@olamy I changed the test to populate the name, but am still getting a test failure on line 192:

https://ci.jenkins.io/blue/organizations/jenkins/Plugins%2Fjira-plugin/detail/PR-213/18/pipeline#step-30-log-139

Any ideas

File: src/test/java/hudson/plugins/jira/JiraCreateIssueNotifierTest.java [hunk 4]
```diff
@@ -168,7 +169,7 @@ public void performFailureSuccessIssueClosedWithComponents() throws Exception {
         doReturn(site).when(notifier).getSiteForProject(Mockito.any());
 
         Issue issue = mock(Issue.class);
-        Status status = new Status(null, null, JiraCreateIssueNotifier.finishedStatuses.Closed.toString() , null, null);
+        Status status = new Status(null, null, "1", JiraCreateIssueNotifier.finishedStatuses.Closed.toString() , null, null);
```
Comment (olamy) [inline src/test/java/hudson/plugins/jira/JiraCreateIssueNotifierTest.java]: name must be one of the member of enum JiraCreateIssueNotifier.finishedStatuses

Comment (rodrigc) [reply src/test/java/hudson/plugins/jira/JiraCreateIssueNotifierTest.java]: Fixed

**Comment by rodrigc** [comment]
@ca-stefan-cordes Thanks for providing all the fixes and collaborating on this.  I wouldn't have been able to solve this without your help!!

**Pull request merged**